distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkActionResponder.java

```diff
@@ -0,0 +1,90 @@
+package com.gruppo1.distributednetworkmanager;
+
+import android.content.Context;
+import android.telephony.SmsMessage;
+
+import com.dezen.riccardo.smshandler.SMSHandler;
+import com.dezen.riccardo.smshandler.SMSManager;
+import com.dezen.riccardo.smshandler.SMSMessage;
+
+public class DistributedNetworkActionResponder {
+
+    private Context context;
+
+
+    public DistributedNetworkActionResponder(Context context){
+
+        this.context = context;
+    }
+
+    /**
+     * This method is called when a message with an action is received
+     *
+     * @param action received action
+     */
+    public void onActionReceived(KadAction action) {
+        switch (action.getActionType()) {
+            case PING:
+                onPingReceived(action);
+                break;
+            case FIND_NODE:
+                onFindNodeReceived(action);
+                break;
+            case FIND_VALUE:
+                onFindValueReceived();
+                break;
+            case STORE:
+                onStoreReceived();
+                break;
+
+                //TODO add a filter for the *_ANSWER action
+        }
+    }
+
+    /**
+     * Method that send the actual response through SMS to the previous sender
+     * @param responseAction
+     */
+    private void sendAction (KadAction responseAction){
+
+        SMSManager.getInstance(context).sendMessage(responseAction.toMessage());
```

Comment:
If you pass an ActionPropagator in the constructor you don't need to send the message through SMSManager, that will be the ActionPropagator's job (most likely the NetworkManager that will keep everything together). For the same reason you don't really need Context in the constructor, you just need NodeDataProvider, ActionPropagator and maybe some kind of interface for Resource access (the NetworkManager will pass a valid one for each of these when instantiating this class)